gobblin-compaction/src/main/java/gobblin/compaction/mapreduce/AbstractDataCompletenessVerifier.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright (C) 2014-2015 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package gobblin.compaction.mapreduce;
+
+import static gobblin.compaction.mapreduce.MRCompactor.FolderState.*;
+
+import gobblin.configuration.State;
+
+
+/**
+ * An abstract class for verifying data completeness of an input folder.
+ *
+ * Subclasses should override the {@link Runnable#run()} method. {@link #afterVerificationStarts()} should be called
+ * right after data completeness verification starts for the input folder, and
+ * {@link #afterVerificationFinishes(boolean)} should be called right after data completeness verification finishes
+ * for the input folder.
+ *
+ * @author ziliu
+ */
+public abstract class AbstractDataCompletenessVerifier
+    implements Runnable, Comparable<AbstractDataCompletenessVerifier> {
+
+  protected final MRCompactor mrCompactor;
+  protected final String inputFolder;
+  protected final State jobProps;
+  protected final double priority;
+
+  public AbstractDataCompletenessVerifier(MRCompactor compactor, String inputFolder, State jobProps, Double priority) {
```

Comment:
I don't think the verifier object should be on a per-folder/dataset level. This precludes the ability to pool multiple completeness checks to the audit backend.

I also think we should decouple the completeness read from the check if the completeness value meets the publish threshold. The former may also be useful for other purposes like monitoring.

I suggest you define a DatasetAuditCountReader interface or something along that lines. It should define a method like ListenableFuture<Long> readAuditCounts(dataset). ListenableFuture allows to add a callback on success or failure which can then modify the state of MRCompactor. Whether there is a threadpool to run the audit count requests there is some kind of elevator algorithm or so should be completely up to the reader implementation. MRCompactor shouldn't be concerned with thread pools and so on.

You can also have a DatasetAuditCountCompletenessChecker which has a checkCompleteness(dataset, count) method to verify the completeness. This should be fairly light-weight and should probably be done synchronously. 

With this model, I don't think you need afterVerificationStarts() and  afterVerificationFinishes(). The process should be driven by MRCompactor and it could provide the necessary callbacks to drive the state transitions.

I also suggest that you add some abstraction around the input folder. I am not sure what this folder is. Is it an HDFS folder? What is the namenode, etc. I think the interfaces should use a dataset interface instead. It can have the attributes that we discussed for the dataset publication. Also consider the use case of running compaction on multiple folders.